sycl/include/CL/sycl/half_type.hpp

```diff
@@ -86,22 +88,132 @@ inline float cast_if_host_half(half_impl::half val) {
 } // namespace sycl
 } // namespace cl
 
+#ifdef __SYCL_DEVICE_ONLY__
+using half = _Float16;
+#else
+using half = cl::sycl::detail::half_impl::half;
+#endif
+
+// Partial specialization of some functions in namespace `std`
 namespace std {
 
-template <> struct hash<cl::sycl::detail::half_impl::half> {
-  size_t operator()(cl::sycl::detail::half_impl::half const &key) const
-      noexcept {
-    return hash<uint16_t>()(key.Buf);
+#ifdef __SYCL_DEVICE_ONLY__
+// `constexpr` could work because the implicit conversion from `float` to
+// `_Float16` can be `constexpr`.
+#define CONSTEXPR_QUALIFIER constexpr
+#else
+// The qualifier is `const` instead of `constexpr` that is original to be
+// because the constructor is not `constexpr` function.
+#define CONSTEXPR_QUALIFIER const
+#endif
+
+// Partial specialization of `std::hash<cl::sycl::half>`
+template <> struct hash<half> {
+  size_t operator()(half const &Key) const noexcept {
+    return hash<uint16_t>()(reinterpret_cast<const uint16_t &>(Key));
   }
 };
 
-} // namespace std
+// Partial specialization of `std::numeric<cl::sycl::half>`
 
-#ifdef __SYCL_DEVICE_ONLY__
-using half = _Float16;
-#else
-using half = cl::sycl::detail::half_impl::half;
-#endif
+// All following values are either calculated based on description of each
+// function/value on https://en.cppreference.com/w/cpp/types/numeric_limits, or
+// cl_platform.h.
+#define HLF_MIN 6.103515625e-05F
```

Comment:
I am wondering whether declaring all these macros increases the risk of conflicts with some user macros...

Reply:
The reason I put them here is in C++ standard library, they have similar macros for these values maybe for the compatibility of C. I intentionally followed the way here. :)

Reply:
May be rename the macros to something like SYCL_HALF_MIN? Such name is better aligned with OpenCL(CL_HALF_MIN) and unique enough to avoid potential conflicts. 
Also there is a macro for radix value(FLT_RADIX for float), perhaps it's analogue also should be added?

Reply:
Make sense. I'll follow up.